This file is used to stratify works and words by year, so that the dataset can be deconstructed

In [7]:
import pandas as pd
import numpy as np
import collections
import pickle
from IPython.display import display

Deserializing DataFrame from basic_dataset.py

In [2]:
words_with_years = pd.read_pickle('./words.pkl')
years_and_books = pd.read_pickle('./books.pkl')
sentences_with_years = pd.read_pickle('./sentences.pkl')
display(words_with_years)
display(years_and_books)
display(sentences_with_years)


0          1         2            3           4             5      \
1838   shining       pale      felt           --   expansion          sole   
1895      pale       felt        --       feudal    tyrannic          dint   
1850      pale       felt    petted           --      feudal       Dashers   
1837    slopes  vermiform      pale         felt          --  warm-blooded   
1831  dripping       pale    'Neath       rustle         Let      Wrapping   
...        ...        ...       ...          ...         ...           ...   
170     Seeing      sects      cava  Hippocratic       Alone   distributed   
1713    Seeing       mite   pursuit        sects        felt            --   
1955    lusted       pale      felt           --         654      scholars   
101       pale       felt  scholars    admitting  courageous         raise   
1660      felt         --       654    admitting      stoics  Opposition.-   

                 6           7              8              9      ... 38586  \
1838               Let   animation    distributed       sidewalk  ...  None   
1895          unlikely       exult        trained           Foul  ...  None   
1850               orb  half-fancy       dilation   unhappiness-  ...  None   
1837         admitting    slowness  entomologists     courageous  ...  None   
1831           Stygian        this          Earth            hid  ...  None   
...                ...         ...            ...            ...  ...   ...   
170          expansion  definitely            Let        regains  ...  None   
1713              sole  imposition            Let       ensnared  ...  None   
1955           unmanly       raise     friendship        sadness  ...  None   
101         friendship     sadness        trained  deliberations  ...  None   
1660  Contradictions.-       raise     friendship        sadness  ...  None   

     38587 38588 38589 38590 38591 38592 38593 38594 38595  
1838  None  None  None  None  None  None  None  None  None  
1895  None  None  None  None  None  None  None  None  None  
1850  None  None  None  None  None  None  None  None  None  
1837  None  None  None  None  None  None  None  None  None  
1831  None  None  None  None  None  None  None  None  None  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...  
170   None  None  None  None  None  None  None  None  None  
1713  None  None  None  None  None  None  None  None  None  
1955  None  None  None  None  None  None  None  None  None  
101   None  None  None  None  None  None  None  None  None  
1660  None  None  None  None  None  None  None  None  None  

[189 rows x 38596 columns]

0                        1   \
1838             poe-ligeia-452.txt  poe-predicament-421.txt   
1895                poe-pit-110.txt              the_pit.txt   
1850            poe-morning-559.txt    poe-landscape-689.txt   
1837            poe-silence-656.txt       poe-sonnet-661.txt   
1831            poe-sleeper-703.txt      poe-israfel-448.txt   
...                             ...                      ...   
170                   faculties.txt                     None   
1713         berkeley-three-745.txt                     None   
1955  augustine-confessions-276.txt                     None   
101    epictetus-discourses-568.txt                     None   
1660         pascal-pensees-569.txt                     None   

                          2                            3   \
1838  hawthorne-lady-472.txt      hawthorne-peter-479.txt   
1895                cask.txt                 telltale.txt   
1850    poe-thousand-708.txt   poe-metzengerstein-557.txt   
1837      poe-bridal-431.txt  hawthorne-prophetic-476.txt   
1831        poe-city-673.txt               poe-to-715.txt   
...                      ...                          ...   
170                     None                         None   
1713                    None                         None   
1955                    None                         None   
101                     None                         None   
1660                    None                         None   

                        4                   5                      6   \
1838                  None                None                   None   
1895          b-p_plan.txt         chilc10.txt  scott-ivanhoe-159.txt   
1850    poe-sphinx-705.txt   poe-three-710.txt        poe-man-691.txt   
1837  hawthorne-dr-468.txt  origin_species.txt                   None   
1831    poe-valley-709.txt  poe-lenore-451.txt                   None   
...                    ...                 ...                    ...   
170                   None                None                   None   
1713                  None                None                   None   
1955                  None                None                   None   
101                   None                None                   None   
1660                  None                None                   None   

                   7                 8                9   ...  \
1838             None              None             None  ...   
1895     redbadge.txt              None             None  ...   
1850  poe-bon-430.txt  poe-some-657.txt  poe-man-690.txt  ...   
1837             None              None             None  ...   
1831             None              None             None  ...   
...               ...               ...              ...  ...   
170              None              None             None  ...   
1713             None              None             None  ...   
1955             None              None             None  ...   
101              None              None             None  ...   
1660             None              None             None  ...   

                    43                     44                45  \
1838              None                   None              None   
1895              None                   None              None   
1850  poe-thou-416.txt  poe-elizabeth-438.txt  poe-loss-455.txt   
1837              None                   None              None   
1831              None                   None              None   
...                ...                    ...               ...   
170               None                   None              None   
1713              None                   None              None   
1955              None                   None              None   
101               None                   None              None   
1660              None                   None              None   

                        46                47             48  \
1838 

0      \
1838                       "Turn him out of the house!"   
1895          The youth leaned heavily upon his friend.   
1850  She answered: "Allah Almighty vouchsafe to the...   
1837  As soon as this occurred, the bees ceased to e...   
1831  Oh, may her sleep       As it is lasting, so b...   
...                                                 ...   
170   The fact, therefore, that Erasistratus, in reg...   
1713  Though indeed I deny they have an existence di...   
1955  Have pity,  O Lord God, lest those who pass by...   
101   have you anything better or greater to see tha...   
1660                                         12-17; 66.   

                                                  1      \
1838      --shall this Conqueror be not once conquered?   
1895                                     -- here, here!   
1850  Now it was this latter peculiarity in his disp...   
1837  Insects often resemble for the sake of protect...   
1831  Sure thou art come O'er far-off seas,       A ...   
...                                                 ...   
170   For, if they do rebound, how then do they pass...   
1713                       So far you are in the right.   
1955                           Why are they not  happy?   
101   I would willingly take a voyage for this purpo...   
1660                                               635.   

                                                  2      \
1838  And in this expectation I was not at all decei...   
1895                      The man screamed: "Let go me!   
1850  It is, indeed, the instinct given to man by Go...   
1837  Modifications in hard parts and in external pa...   
1831                Heaven have her in its sacred keep!   
...                                                 ...   
170   Now, since the three faculties of Nature have ...   
1713  In short, by whatever method you distinguish <...   
1955                                    [633]       37.   
101   And that you may not think that I show you the...   
1660  What mysteries has he taught, even according t...   

                                                  3      \
1838  Amazement now struggled in my bosom with the p...   
1895  Prince  John  coloured  as  he  put  the quest...   
1850  observed the Baron, dryly, and at that instant...   
1837  Secondary Sexual Characters Variable.- I think...   
1831  Nothing there is motionless-        Nothing sa...   
...                                                 ...   
170             *What we now call the pulmonary artery.   
1713  And yet you asserted that you could not concei...   
1955  [297]  But what is like to  thy Word, our Lord...   
101   CHAPTER 5    How magnanimity is consistent wit...   
1660                                           Himself.   

                                                  4      \
1838  "Did not my great-granduncle, Peter Goldthwait...   
1895  We came at length to the foot of the descent, ...   
1850  For even as thou entreatedst me generously, wi...   
1837  Drift timber is thrown up on most islands, eve...   
1831  There open fanes and gaping graves       Yawn ...   
...                                                 ...   
170                        For this simple vessel [i.e.   
1713  It cannot be denied that we perceive such cert...   
1955                              CHAPTER VII       11.   
101             for having acted conformably to nature?   
1660  Nature sustains our feeble reason and prevents...   

                                                  5      \
1838  What a host of gloomy recollections will ever ...   
1895                 and look to it what thou wilt do!"   
1850                                        LA BRUYERE.   
1837  LARYNX, The upper part of the windpipe opening...   
1831                                        -THE END- .   
...                                                 ...   
170                                        the tissues.   
1713  The colours, therefore, by 

Sorting each row by year in order to create proper time period bins.

In [3]:
words_with_years = words_with_years.sort_index()
display(words_with_years)
years_and_books = years_and_books.sort_index()
display(years_and_books)
sentences_with_years = sentences_with_years.sort_index()
display(sentences_with_years)

0           1         2             3           4           5      \
50      slopes    dripping      felt           orb       raise  friendship   
59     pursuit     shining      felt            --         Let   Tentmaker   
64        felt  imposition      sole         well-    Britains     provide   
87        pale        felt        --   garrisoning        Sept   admitting   
101       pale        felt  scholars     admitting  courageous       raise   
...        ...         ...       ...           ...         ...         ...   
1991      pale        felt        --  serpent-kind         orb      Hasted   
1992  dripping        pale      felt            --      petted         ord   
1993   catkins     brained      pale          felt          --      petted   
1994      pale      petted        --          felt    tyrannic         orb   
2001      pale        felt    petted            --         orb        dint   

               6         7            8              9      ... 38586  \
50              brim   trained        finds       savagely  ...  None   
59            Kheyam   revisit        Earth        provide  ...  None   
64        structures      make      dangers      onslaught  ...  None   
87             extra     Adams            l          raise  ...  None   
101       friendship   sadness      trained  deliberations  ...  None   
...              ...       ...          ...            ...  ...   ...   
1991            dint     raise      disport       repealed  ...  None   
1992       admitting     Adams        raise     friendship  ...  None   
1993          feudal       orb  silhouettes   lifelessness  ...  Past   
1994   Chattahoochee   unmanly         dint       repealed  ...  None   
2001  inconsequently  repealed     bordered       unlikely  ...  None   

         38587    38588 38589 38590       38591        38592     38593 38594  \
50        None     None  None  None        None         None      None  None   
59        None     None  None  None        None         None      None  None   
64        None     None  None  None        None         None      None  None   
87        None     None  None  None        None         None      None  None   
101       None     None  None  None        None         None      None  None   
...        ...      ...   ...   ...         ...          ...       ...   ...   
1991      None     None  None  None        None         None      None  None   
1992      None     None  None  None        None         None      None  None   
1993  infamous  unknown  code  duty  recuperate  Wonderingly  hillside   And   
1994      None     None  None  None        None         None      None  None   
2001      None     None  None  None        None         None      None  None   

     38595  
50    None  
59    None  
64    None  
87    None  
101   None  
...    ...  
1991  None  
1992  None  
1993   XIX  
1994  None  
2001  None  

[189 rows x 38596 columns]

0                          1   \
50            lucretius-on-395.txt              on_nature.txt   
59                     rubai10.txt                       None   
64                    politics.txt                       None   
87         franklin-london-246.txt                       None   
101   epictetus-discourses-568.txt                       None   
...                            ...                        ...   
1991       milton-paradise-108.txt                       None   
1992                  aesopa10.txt                opion11.txt   
1993                   tarz210.txt      burroughs-son-333.txt   
1994          burroughs-at-316.txt  burroughs-monster-323.txt   
2001                   pmars10.txt                 anne11.txt   

                            2                3                         4   \
50                        None             None                      None   
59                        None             None                      None   
64                        None             None                      None   
87                        None             None                      None   
101                       None             None                      None   
...                        ...              ...                       ...   
1991                      None             None                      None   
1992               hydea10.txt             None                      None   
1993  burroughs-return-326.txt      tarz310.txt  burroughs-beasts-317.txt   
1994  burroughs-jungle-321.txt  twain-30-44.txt  stevenson-childs-163.txt   
2001               scarp10.txt       song10.txt               carol10.txt   

                              5                         6   \
50                          None                      None   
59                          None                      None   
64                          None                      None   
87                          None                      None   
101                         None                      None   
...                          ...                       ...   
1991                        None                      None   
1992                        None                      None   
1993   burroughs-warlord-364.txt  burroughs-tarzan-334.txt   
1994  stevenson-treasure-166.txt       austen-emma-754.txt   
2001                  alad10.txt                avon10.txt   

                            7                     8                       9   \
50                        None                  None                    None   
59                        None                  None                    None   
64                        None                  None                    None   
87                        None                  None                    None   
101                       None                  None                    None   
...                        ...                   ...                     ...   
1991                      None                  None                    None   
1992                      None                  None                    None   
1993  burroughs-thuvia-336.txt           yanke10.txt             yanke11.txt   
1994        hardy-jude-132.txt  burnett-sara-308.txt  burnett-secret-313.txt   
2001              scarlet1.txt          plrabn12.txt                    None   

      ...    43    44    45    46    47    48    49    50    51    52  
50    ...  None  None  None  None  None  None  None  None  None  None  
59    ...  None  None  None  None  None  None  None  None  None  None  
64    ...  None  None  None  None  None  None  None  None  None  None  
87    ...  None  None  None  None  None  None  None  None  None  None  
101   ...  None  None  None  None  None  None  None  None  None  None  
...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
1991  ...  None  None  None  None  None  None  None  None  None  None  
1992  ...  None  None  None  None  None  Non

0      \
50    And commonly   The awnings, saffron, red and d...   
59    That ev'n my buried Ashes such a Snare  Of Per...   
64                            It was an apt comparison.   
87    And we shall continue to enjoy, _in fact_, the...   
101   have you anything better or greater to see tha...   
...                                                 ...   
1991  Now had night measured with her shadowy cone  ...   
1992                                "No guns, no guns!"   
1993                       I want to go on living HERE.   
1994  But I was too quick for it, and so, half hoppi...   
2001  Now had night measured with her shadowy cone  ...   

                                                  1      \
50    Grant they could,   Yet by their meetings and ...   
59    I tell Thee this--When, starting from the Goal...   
64    WORKERS RESEARCH UNIT, "BELFAST BULLETIN      ...   
87    1775          LETTERS  REASONS AGAINST SATIRIZ...   
101   I would willingly take a voyage for this purpo...   
...                                                 ...   
1991  Regions they passed, the mighty regencies  Of ...   
1992  But since she had been here the sounds and sme...   
1993  Tarzan smiled, and then, bowing to the count, ...   
1994  Scraping the bristles from the pig's upturned ...   
2001  Mr. Larsen was a small, plump man, with a shor...   

                                                  2      \
50    So sayst thou; well, huge many a river seems  ...   
59    Into this Universe, and Why not knowing  Nor W...   
64    "NOTES ON EMERGENCY LEGISLATION"           FAL...   
87    They all declare against the Church, and again...   
101   And that you may not think that I show you the...   
...                                                 ...   
1991  Round he surveys (and well might, where he sto...   
1992                             Emil went to her door.   
1993  The young Arab pretended to do as he was bid, ...   
1994  However, here he was; and in fulfilment of his...   
2001  Marguerite's heart seemed all at once to stop ...   

                                                  3      \
50    But knew I never what   The seeds primordial w...   
59    ], "When   Nizam-ul-Mulk was in the Agony (of ...   
64    At the time of her killing, the RUC claimed to...   
87    You say, poor labourers cannot afford to buy b...   
101   CHAPTER 5    How magnanimity is consistent wit...   
...                                                 ...   
1991  First crept  The parsimonious emmet, provident...   
1992  If you had any nerve you'd get together and ma...   
1993             "Thurid, Dator of Issus, has no price.   
1994  That there's a picture of a missel thrush on h...   
2001  However, she reminded Davy that he ought to sa...   

                                                  4      \
50    The sojourners in the Stymphalian fens   So dr...   
59                                       in the Desert.   
64    Hillsborough marked a new development in count...   
87                      They will petition for Redress.   
101             for having acted conformably to nature?   
...                                                 ...   
1991  Remember with what mild  And gracious temper h...   
1992  Under a white mulberry tree there was an old w...   
1993         Tarzan arose lazily and stretched himself.   
1994  Its tendency would be to raise and refine her ...   
2001  "I met him on the playground, such a dripping ...   

                                                  5      \
50    But ocean uprisen would often rave in vain,   ...   
59                                                  IX.   
64    This strategy had as its objectives the crimin...   
87    Convert the brave honest Officers of your Navy...   
101   For philosophers say we allow none to be free ...   
...                                                 ...   
1991      To whom thus Adam, cleared of doubt, replied.   
1992  The Tortoise and the Birds 

Finding evenly ditributed time period bins, by number of works in those years, and stratifying accordingly.

In [4]:
year_freqs = []
total_sum = 0
for row in years_and_books.iterrows():
    freq = 0
    for elem in row[1].iteritems():
        if elem[1] is None:
            year_freqs.append([row[0], freq])
            break
        else:
            freq += 1
            total_sum += 1
current_bin = 1
num_bins = 4
bins = []
ind = 0
start = 0
end = 0
temp_sum = 0
while ind < len(year_freqs)-1:
    temp_sum += year_freqs[ind][1]
    if temp_sum > current_bin*total_sum/num_bins:
        end = ind
        bins.append([year_freqs[start][0], year_freqs[end][0]])
        start = ind+1
        current_bin += 1
    ind += 1
bins.append([year_freqs[start][0], year_freqs[ind][0]])
print(bins)






        

[[50, 1775], [1776, 1841], [1842, 1898], [1899, 2001]]


These 4 bins most closely correspond to existing eras of literature, those being:

Classical/Neoclassical: before 1660, 1660-1790
Romantic: 1790-1830
Victorian: 1932-1901
Modern/Postmodern: 1914-Present

Source: https://www.summitlearning.org/docs/227310

Now that we have our bins, we can separate the existing dataset into these 4 categories.

In [5]:
classical_b = years_and_books.loc[bins[0][0]:bins[0][1]]
classical_w = words_with_years.loc[bins[0][0]:bins[0][1]]
classical_s = sentences_with_years.loc[bins[0][0]:bins[0][1]]

romantic_b = years_and_books.loc[bins[1][0]:bins[1][1]]
romantic_w = words_with_years.loc[bins[1][0]:bins[1][1]]
romantic_s = sentences_with_years.loc[bins[1][0]:bins[1][1]]

victorian_b = years_and_books.loc[bins[2][0]:bins[2][1]]
victorian_w = words_with_years.loc[bins[2][0]:bins[2][1]]
victorian_s = sentences_with_years.loc[bins[2][0]:bins[2][1]]

modern_b = years_and_books.loc[bins[3][0]:bins[3][1]]
modern_w = words_with_years.loc[bins[3][0]:bins[3][1]]
modern_s = sentences_with_years.loc[bins[3][0]:bins[3][1]]


Then we separate these into NumPy arrays to consolidate them into a single dataset of books and words according to the era.

In [6]:
classical_bs = classical_b.to_numpy()
classical_ws = classical_w.to_numpy()
classical_ss = classical_s.to_numpy()

romantic_bs = romantic_b.to_numpy()
romantic_ws = romantic_w.to_numpy()
romantic_ss = romantic_s.to_numpy()

victorian_bs = victorian_b.to_numpy()
victorian_ws = victorian_w.to_numpy()
victorian_ss = victorian_s.to_numpy()

modern_bs = modern_b.to_numpy()
modern_ws = modern_w.to_numpy()
modern_ss = modern_s.to_numpy()

#final classical dataset
classical = collections.defaultdict()
class_b = np.array([])
for b in classical_bs:
    b = [x for x in b if x is not None]
    class_b = np.concatenate((class_b, b))
class_w = np.array([])
for w in classical_ws:
    w = [x for x in w if x is not None]
    class_w = np.concatenate((class_w, w))
class_w = np.unique(class_w)
class_s = np.array([])
for s in classical_ss:
    s = [x for x in w if s is not None]
    class_s = np.concatenate((class_s, s))
    
classical['words'] = class_w
classical['books'] = class_b
classical['sentences'] = class_s

#final romantic dataset
romantic = collections.defaultdict()
rom_b = np.array([])
for b in romantic_bs:
    b = [x for x in b if x is not None]
    rom_b = np.concatenate((rom_b, b))
rom_w = np.array([])
for w in romantic_ws:
    w = [x for x in w if x is not None]
    rom_w = np.concatenate((rom_w, w))
rom_w = np.unique(rom_w)
rom_s = np.array([])
for s in romantic_ss:
    s = [x for x in s if x is not None]
    rom_s = np.concatenate((rom_s, s))
    
romantic['words'] = rom_w
romantic['books'] = rom_b
romantic['sentences'] = rom_s

#final victorian dataset
victorian = collections.defaultdict()
vic_b = np.array([])
for b in victorian_bs:
    b = [x for x in b if x is not None]
    vic_b = np.concatenate((vic_b, b))
vic_w = np.array([])
for w in victorian_ws:
    w = [x for x in w if x is not None]
    vic_w = np.concatenate((vic_w, w))
vic_w = np.unique(vic_w)
vic_s = np.array([])
for s in victorian_ss:
    s = [x for x in s if x is not None]
    vic_s = np.concatenate((vic_s, s))
    
victorian['words'] = vic_w
victorian['books'] = vic_b
victorian['sentences'] = vic_s

#final modern dataset
modern = collections.defaultdict()
mod_b = np.array([])
for b in modern_bs:
    b = [x for x in b if x is not None]
    mod_b = np.concatenate((mod_b, b))
mod_w = np.array([])
for w in modern_ws:
    w = [x for x in w if x is not None]
    mod_w = np.concatenate((mod_w, w))
mod_w = np.unique(mod_w)
mod_s = np.array([])
for s in modern_ss:
    s = [x for x in s if x is not None]
    mod_s = np.concatenate((mod_s, s))
    
modern['words'] = mod_w
modern['books'] = mod_b
modern['sentences'] = mod_s

['Is he dead?'
 'The cowboy and the Easterner burnished themselves fiery red with this water, until it seemed to be some kind of a metal polish.'
 '"I keep a hotel," he shouted.' ...
 '"Two o\'clock--but isn\'t it splendid about the picnic, Marilla?'
 'He did not speak or move, but the blotchy green of his countenance turned livid, and the froth froze upon his lips.'
 'He took off his soutane, beneath which he was wearing close-fitting breeches and a cloth waistcoat, and began changing his attire.']


Now, with the final dataset assembled, with each era's dictionary containing the corresponding unique words, sentences, and books, we can serialize our 4 datasets so we can create the word embeddings in the next file.

In [8]:
pickle.dump(classical, open('classical.pkl', 'wb'), protocol=pickle.HIGHEST_PROTOCOL)
pickle.dump(romantic, open('romantic.pkl', 'wb'), protocol=pickle.HIGHEST_PROTOCOL)
pickle.dump(victorian, open('victorian.pkl', 'wb'), protocol=pickle.HIGHEST_PROTOCOL)
pickle.dump(modern, open('modern.pkl', 'wb'), protocol=pickle.HIGHEST_PROTOCOL)